### PR TITLE
[Merged by Bors] - feat(Combinatorics/SimpleGraph): lemmas for IsMatching

### PR DESCRIPTION
**Pull request by pimotte**
Supporting lemmas for deducing that a subgraph is a matching.
Added in preparation for Tutte's theorem.

---

- [x] depends on: #11667 

This is in preparation for Tutte's theorem. My (partially complete) approach can be found [here](https://github.com/pimotte/TutteLean/blob/main/TutteLean.lean)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean [hunk 1]
```diff
@@ -981,6 +983,17 @@
   exact fun _ _ ↦ False.elim
 #align simple_graph.singleton_subgraph_snd_le_subgraph_of_adj SimpleGraph.singletonSubgraph_snd_le_subgraphOfAdj
```
[the remark this answers is not in the document]
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:984]: Fixed in b6f654f64ca80471a2f7c56223bdb65288c68bcc

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 2]
```diff
@@ -36,70 +36,149 @@ one edge, and the edges of the subgraph represent the paired vertices.
...
+  obtain ⟨ w , hw ⟩ := hM (Set.mem_of_mem_image_val ((Subgraph.coeSubgraph_verts M).symm ▸ hv))
+  use w
+  constructor
+  · simp_rw [Subgraph.coeSubgraph_adj]
+    simp only [Subtype.forall, Subtype.coe_eta, Subtype.coe_prop, exists_const] at *
+    obtain ⟨ v' , hv' ⟩ := (Set.mem_image _ _ _).mp ((Subgraph.coeSubgraph_verts M).symm ▸ hv)
+    exact ⟨hv'.2 ▸ v'.2, hw.1⟩
+  · intro y hy
+    obtain ⟨ hv' , ⟨ hw' , hvw ⟩ ⟩ := (SimpleGraph.Subgraph.coeSubgraph_adj _ _ _).mp hy
+    rw [← hw.2 ⟨ y , hw' ⟩ hvw]
```
Comment (Rida-Hamadani) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:161]:
```suggestion
  intro _ hv
  obtain ⟨w, hw⟩ := hM <| Set.mem_of_mem_image_val <| M.coeSubgraph_verts.symm ▸ hv
  use w
  refine ⟨?_, fun y hy => ?_⟩
  · obtain ⟨v, hv⟩ := (Set.mem_image _ _ _).mp <| (coeSubgraph_verts M).symm ▸ hv
    simp only [coeSubgraph_adj, Subtype.coe_eta, Subtype.coe_prop, exists_const]
    exact ⟨hv.2 ▸ v.2, hw.1⟩
  · obtain ⟨_, hw', hvw⟩ := (coeSubgraph_adj _ _ _).mp hy
    rw [← hw.2 ⟨y, hw'⟩ hvw]
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 3]
```diff
@@ -119,17 +198,75 @@ theorem isPerfectMatching_iff : M.IsPerfectMatching ↔ ∀ v, ∃! w, M.Adj v w
...
+lemma IsMatching.induce_connectedComponent (h : M.IsMatching) (c : ConnectedComponent G) :
+    (M.induce (M.verts ∩ c.supp)).IsMatching := by
+  intro v hv
+  simp only [induce_verts, Set.mem_inter_iff, ConnectedComponent.mem_supp_iff] at hv
+  obtain ⟨hv, rfl⟩ := hv
+  obtain ⟨w, hvw, hw⟩ := h hv
+  use w
+  simp only [induce_adj, Set.mem_inter_iff, hv, ConnectedComponent.mem_supp_iff, and_self, and_imp,
+    true_and, ConnectedComponent.eq, hw, hvw, M.edge_vert hvw.symm, (M.adj_sub hvw).symm.reachable]
+  exact fun _ _ _ ↦ hw _
```
Comment (Rida-Hamadani) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:221]:
```suggestion
intro _ hv
  obtain ⟨hv, rfl⟩ := hv
  obtain ⟨w, hvw, hw⟩ := h hv
  use w
  simpa [hv, hvw, M.edge_vert hvw.symm, (M.adj_sub hvw).symm.reachable] using fun _ _ _ ↦ hw _
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 4]
```diff
@@ -119,17 +198,75 @@ theorem isPerfectMatching_iff : M.IsPerfectMatching ↔ ∀ v, ∃! w, M.Adj v w
...
 
+namespace ConnectedComponent
+
+section Finite
+
+variable [Fintype V] [DecidableEq V] [DecidableRel G.Adj]
+
+instance instDecidableMemSupp (c : G.ConnectedComponent) (v : V) : Decidable (v ∈ c.supp) :=
+  c.recOn (fun w ↦ decidable_of_iff (G.Reachable v w) $ by simp)
+    (fun _ _ _ _ ↦ Subsingleton.elim _ _)
```
Comment (Rida-Hamadani) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:238]:
```suggestion
  c.recOn fun w ↦ decidable_of_iff (G.Reachable v w) <| by simp fun _ _ _ _ ↦ Subsingleton.elim _ _
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 5]
```diff
@@ -119,17 +198,75 @@ theorem isPerfectMatching_iff : M.IsPerfectMatching ↔ ∀ v, ∃! w, M.Adj v w
...
+        and_true] at hv' ⊢
+      trivial
+
+    apply Nat.odd_iff_not_even.mp codd
+    haveI : Fintype ↑(Subgraph.induce M (Subtype.val '' supp c)).verts := Fintype.ofFinite _
+    have hMeven := Subgraph.IsMatching.even_card hMmatch
+    haveI : Fintype (c.supp) := Fintype.ofFinite _
+    simp only [Subgraph.induce_verts, Subgraph.verts_top, Set.toFinset_image,
+      Nat.card_eq_fintype_card, Set.toFinset_image,
+      Finset.card_image_of_injective _ (Subtype.val_injective), Set.toFinset_card] at hMeven ⊢
```
Comment (Rida-Hamadani) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:267]:
```suggestion
      obtain ⟨⟨v', hv'⟩, hv , rfl⟩ := hv
      use w
      have hwnu : w ∉ u := fun hw' ↦ h w hw' ⟨v', hv'⟩ hw.1 hv
      refine ⟨⟨⟨⟨v', hv'⟩, hv, rfl⟩, ?_, hw.1⟩, fun _ hy ↦ hw.2 _ hy.2.2⟩
      apply mem_coe_supp_of_adj ⟨⟨v', hv'⟩, ⟨hv, rfl⟩⟩ ⟨by trivial, hwnu⟩
      simp only [Subgraph.induce_verts, Subgraph.verts_top, Set.mem_diff, Set.mem_univ, true_and,
        Subgraph.induce_adj, Subgraph.top_adj, M.adj_sub hw.1] at hv' ⊢
      trivial
    apply Nat.odd_iff_not_even.mp codd
    haveI : Fintype (Subgraph.induce M (Subtype.val '' c.supp)).verts := Fintype.ofFinite _
    haveI : Fintype c.supp := Fintype.ofFinite _
    have hMeven := Subgraph.IsMatching.even_card hMmatch
    simp only [Subgraph.induce_verts, Set.toFinset_image, Nat.card_eq_fintype_card,
      Set.toFinset_card, Finset.card_image_of_injective _ (Subtype.val_injective)] at hMeven ⊢
```

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean [hunk 6]
```diff
@@ -981,6 +983,17 @@ theorem singletonSubgraph_snd_le_subgraphOfAdj {u v : V} {h : G.Adj u v} :
...
+lemma subgraphOfAdj_support {u v : V}
+    (h : G.Adj u v) : (G.subgraphOfAdj h).support = {u , v} := by
+  ext v'
+  rw [SimpleGraph.Subgraph.mem_support]
+  simp only [Set.mem_insert_iff, Set.mem_singleton_iff,
+    subgraphOfAdj_adj, Sym2.eq, Sym2.rel_iff', Prod.mk.injEq, Prod.swap_prod_mk]
+  constructor
+  · rintro ⟨ w , hw ⟩
+    exact hw.elim (fun h1 ↦ .inl h1.1.symm) (fun hr ↦ .inr hr.2.symm)
+  · exact fun hv' ↦ hv'.elim (fun hl ↦ ⟨v, .inl ⟨hl.symm, rfl⟩⟩) (fun hr ↦ ⟨u, .inr ⟨rfl, hr.symm⟩⟩)
```
Comment (Rida-Hamadani) [inline Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:995]:
```suggestion
lemma subgraphOfAdj_support {u v : V} (h : G.Adj u v) :
    (G.subgraphOfAdj h).support = {u , v} := by
  ext
  rw [Subgraph.mem_support]
  simp only [subgraphOfAdj_adj, Sym2.eq, Sym2.rel_iff', Prod.mk.injEq, Prod.swap_prod_mk]
  refine ⟨?_, fun h ↦ h.elim (fun hl ↦ ⟨v, .inl ⟨hl.symm, rfl⟩⟩) fun hr ↦ ⟨u, .inr ⟨rfl, hr.symm⟩⟩⟩
  rintro ⟨w, hw⟩
  exact hw.elim (fun h1 ↦ .inl h1.1.symm) fun hr ↦ .inr hr.2.symm
```

**Review by Rida-Hamadani — COMMENTED** [review]
Style and shorten.

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 7]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
   rw [h.toEdge_eq_of_adj hv ha, h.toEdge_eq_of_adj hw (M.symm ha), Subtype.mk_eq_mk, Sym2.eq_swap]
 #align simple_graph.subgraph.is_matching.to_edge_eq_to_edge_of_adj SimpleGraph.Subgraph.IsMatching.toEdge_eq_toEdge_of_adj
 
+lemma IsMatching.map_hom_ofLE_IsMatching (h : M.IsMatching) (hGG' : G ≤ G') :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:79]: To match the naming convention:
```suggestion
lemma IsMatching.map_ofLE (h : M.IsMatching) (hGG' : G ≤ G') :
```
or
```suggestion
lemma IsMatching.map_homOfLE (h : M.IsMatching) (hGG' : G ≤ G') :
```
(I think the first one is better)

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 8]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
...
+    (M.map (Hom.ofLE hGG')).IsMatching := by
+  intro _ hv
+  obtain ⟨_, hv, hv'⟩ := Set.mem_image _ _ _ |>.mp hv
+  obtain ⟨w, hw⟩ := h hv
+  use w
+  simpa using hv' ▸ hw
+
+lemma sup_IsMatching (hM : M.IsMatching) (hM' : M'.IsMatching)
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:87]:
```suggestion
lemma IsMatching.sup (hM : M.IsMatching) (hM' : M'.IsMatching)
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 9]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
...
+    | inr h =>
+      rw [Set.disjoint_left] at hd
+      simpa [(mem_support _).mpr ⟨w, hw.1⟩, (mem_support _).mpr ⟨y, h⟩] using @hd v
+  cases Set.mem_or_mem_of_mem_union hv with
+  | inl hmM => exact aux hM hd hmM
+  | inr hmM' =>
+    rw [sup_comm]
+    exact aux hM' (Disjoint.symm hd) hmM'
+
+lemma iSup_IsMatching {ι : Sort _} {f : ι → Subgraph G} (hM : (i : ι) → (f i).IsMatching)
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:107]:
```suggestion
lemma IsMatching.iSup {ι : Sort _} {f : ι → Subgraph G} (hM : (i : ι) → (f i).IsMatching)
```
for *anonymous dot notation*: This lets you write `.iSup _ : IsMatching _`

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean [hunk 10]
```diff
@@ -979,6 +979,15 @@ theorem singletonSubgraph_snd_le_subgraphOfAdj {u v : V} {h : G.Adj u v} :
   simp
 #align simple_graph.singleton_subgraph_snd_le_subgraph_of_adj SimpleGraph.singletonSubgraph_snd_le_subgraphOfAdj
 
+lemma subgraphOfAdj_support {u v : V} (h : G.Adj u v) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:982]: The use of dot notation shouldn't change the lemma name
```suggestion
lemma support_subgraphOfAdj {u v : V} (h : G.Adj u v) :
```

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean [hunk 11]
```diff
@@ -999,6 +1008,9 @@ protected abbrev restrict {G' : G.Subgraph} : G.Subgraph → G'.coe.Subgraph :=
   Subgraph.comap G'.hom
 #align simple_graph.subgraph.restrict SimpleGraph.Subgraph.restrict
 
+lemma coeSubgraph_verts {G' : Subgraph G} (G'' : Subgraph G'.coe) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:1011]: Same here
```suggestion
lemma verts_coeSubgraph {G' : Subgraph G} (G'' : Subgraph G'.coe) :
```

Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:1011]: Simp lemma, no?

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 12]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
...
+    simpa [(mem_support _).mpr ⟨w, hw.1⟩, (mem_support _).mpr ⟨y, hi'⟩] using @this v
+
+lemma subgraphOfAdj_IsMatching (h : G.Adj v w) : (G.subgraphOfAdj h).IsMatching := by
+  intro _ hv
+  rw [subgraphOfAdj_verts, Set.mem_insert_iff, Set.mem_singleton_iff] at hv
+  cases hv with
+  | inl => use w; aesop
+  | inr => use v; aesop
+
+lemma coe_IsMatching {G' : Subgraph G} {M : Subgraph G'.coe} (hM : M.IsMatching) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:130]:
```suggestion
lemma IsMatching.coeSubgraph {G' : Subgraph G} {M : Subgraph G'.coe} (hM : M.IsMatching) :
```

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 13]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
...
+  refine ⟨iSup_adj.mpr ⟨i, hw.1⟩, ?_⟩
+  intro y hy
+  obtain ⟨i' , hi'⟩ := iSup_adj.mp hy
+  by_cases heq : i = i'
+  · exact hw.2 y (heq.symm ▸ hi')
+  · have := hd _ _ heq
+    rw [Set.disjoint_left] at this
+    simpa [(mem_support _).mpr ⟨w, hw.1⟩, (mem_support _).mpr ⟨y, hi'⟩] using @this v
+
+lemma subgraphOfAdj_IsMatching (h : G.Adj v w) : (G.subgraphOfAdj h).IsMatching := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:123]:
```suggestion
lemma IsMatching.subgraphOfAdj (h : G.Adj v w) : (G.subgraphOfAdj h).IsMatching := by
```
Hopefully you get the pattern from these

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean (hunk 11 above)
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Subgraph.lean]: Added

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 14]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
...
+      rw [Set.disjoint_left] at hd
+      simpa [(mem_support _).mpr ⟨w, hw.1⟩, (mem_support _).mpr ⟨y, h⟩] using @hd v
+  cases Set.mem_or_mem_of_mem_union hv with
+  | inl hmM => exact aux hM hd hmM
+  | inr hmM' =>
+    rw [sup_comm]
+    exact aux hM' (Disjoint.symm hd) hmM'
+
+lemma IsMatching.iSup {ι : Sort _} {f : ι → Subgraph G} (hM : (i : ι) → (f i).IsMatching)
+    (hd : (i j : ι) → (i ≠ j) →  Disjoint ((f i).support) ((f j).support)) :
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:108]:
```suggestion
    (hd : Pairwise fun i j ↦ Disjoint (f i).support (f j).support) :
```

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean [hunk 15]
```diff
@@ -979,6 +979,15 @@ theorem singletonSubgraph_snd_le_subgraphOfAdj {u v : V} {h : G.Adj u v} :
   simp
 #align simple_graph.singleton_subgraph_snd_le_subgraph_of_adj SimpleGraph.singletonSubgraph_snd_le_subgraphOfAdj
 
+lemma support_subgraphOfAdj {u v : V} (h : G.Adj u v) :
+    (G.subgraphOfAdj h).support = {u , v} := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:983]: Probably good as a simp lemma?

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean [hunk 16]
```diff
@@ -979,6 +979,15 @@ theorem singletonSubgraph_snd_le_subgraphOfAdj {u v : V} {h : G.Adj u v} :
   simp
 #align simple_graph.singleton_subgraph_snd_le_subgraph_of_adj SimpleGraph.singletonSubgraph_snd_le_subgraphOfAdj
```
Comment (pimotte) [inline Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:981]:
```suggestion

@[simp]
```

File: Mathlib/Combinatorics/SimpleGraph/Subgraph.lean (hunk 15 above)
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Subgraph.lean:984]: Added:)

File: Mathlib/Combinatorics/SimpleGraph/Matching.lean [hunk 17]
```diff
@@ -76,6 +76,69 @@ theorem IsMatching.toEdge_eq_toEdge_of_adj (h : M.IsMatching)
...
+      simpa [(mem_support _).mpr ⟨w, hw.1⟩, (mem_support _).mpr ⟨y, h⟩] using @hd v
+  cases Set.mem_or_mem_of_mem_union hv with
+  | inl hmM => exact aux hM hd hmM
+  | inr hmM' =>
+    rw [sup_comm]
+    exact aux hM' (Disjoint.symm hd) hmM'
+
+lemma IsMatching.iSup {ι : Sort _} {f : ι → Subgraph G} (hM : (i : ι) → (f i).IsMatching)
+    (hd : Pairwise fun i j ↦ Disjoint (f i).support (f j).support) :
+    (⨆ i , f i).IsMatching := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Matching.lean:109]:
```suggestion
    (⨆ i, f i).IsMatching := by
```

**Review by YaelDillies — APPROVED** [review]
Nice now!

maintainer merge

**Comment by riccardobrasca** [comment]
Taking into account Yaël comments LGTM, thanks!

bors d+

**Pull request closed without merging**